File: lib/utils/font.dart
const weights = {
  100: 'thin',
  200: 'extraLight',
  300: 'light',
  400: 'regular',
  500: 'medium',
  600: 'semiBold',
  700: 'bold',
  800: 'extraBold',
  900: 'heavy',
  950: 'extraBlack',
};

const fallbacks = ['Symbola', 'Noto-CJK'];

typedef Metadata = ({String? name, String style, int weight, String type});

class Font {
  final String path;
  final String fileName;
  final String name;
  final String type;
  final int weight;
  final String style;
  late final List<String> fallbacks;

  get fileNameNoExtension => fileName.replaceAll(RegExp(r'(\.(otf|ttf))$'), '');

  Font({
    required this.path,
    required this.fileName,
    required this.type,
    required this.weight,
    required this.style,
    required Metadata metadata,
  }) : name = metadata.name ?? fileName {
    fallbacks = fallbacks.where((f) => f != fileNameNoExtension).toList();
  }
}

File: website/src/utils/font.ts
import { GlobalFonts } from "@napi-rs/canvas";

const weights = {
  100: "thin",
  200: "extraLight",
  300: "light",
  400: "normal",
  500: "medium",
  600: "semiBold",
  700: "bold",
  800: "extraBold",
  900: "heavy",
  950: "extraBlack",
};
const fallbacks = ["Symbola", "Noto-CJK"];

export default class Font {
  public name: string;
  public style: string;
  public weight: string;
  public type: string;
  public registered: boolean;
  public fallbacks: string[];
  public constructor(
    public path: string,
    public filename: string,
    public metadata: {
      style: string;
      weight: number | string;
      type: string;
      name: string;
    },
  ) {
    this.name = metadata.name || filename;
    this.style = metadata.style === "regular"
      ? "normal"
      : metadata.style || "normal";
    //@ts-expect-error
    this.weight = weights[metadata.weight] || metadata.weight || "normal";
    this.type = metadata.type;
    this.registered = false;
    this.fallbacks = fallbacks.filter((fallback) =>
      fallback !== this.filenameNoExt
    );
  }

  register() {
    if (this.registered) return null;
    this.registered = true;
    return GlobalFonts.registerFromPath(this.path, this.filenameNoExt);
  }

  public toCanvasString(size: number, shouldDoFallback = true) {
    const shouldFall = shouldDoFallback ? `, ${this.fallbacks.join(", ")}` : "";

    // Return the font string, for skia-canvas
    return `${this.weight} ${this.style} ${size}px ${this.filenameNoExt}${shouldFall}`
  }

  get filenameNoExt() {
    return this.filename.replace(/(\.(otf|ttf))$/, "");
  }

  get isFallback() {
    return fallbacks.includes(this.filenameNoExt);
  }
}
